diff -rc ruby-2.1.0/thread_pthread.c ruby-2.1.0-haiku/thread_pthread.c
*** ruby-2.1.0/thread_pthread.c	2013-11-18 22:47:12.000000000 +0900
--- ruby-2.1.0-haiku/thread_pthread.c	2014-05-28 09:31:49.076631932 +0900
***************
*** 34,39 ****
--- 34,43 ----
  #include <sys/time.h>
  #endif
  
+ #if defined(__HAIKU__)
+ #include <kernel/OS.h>
+ #endif
+ 
  static void native_mutex_lock(pthread_mutex_t *lock);
  static void native_mutex_unlock(pthread_mutex_t *lock);
  static int native_mutex_trylock(pthread_mutex_t *lock);
***************
*** 537,542 ****
--- 541,548 ----
      return 0;
  }
  #define pthread_attr_getstackaddr(attr, addr) hpux_attr_getstackaddr(attr, addr)
+ #elif defined __HAIKU__
+ #define STACKADDR_AVAILABLE 1
  #endif
  
  #ifndef MAINSTACKADDR_AVAILABLE
***************
*** 610,615 ****
--- 616,626 ----
      *addr = thinfo.__pi_stackaddr;
      *size = thinfo.__pi_stacksize;
      STACK_DIR_UPPER((void)0, (void)(*addr = (char *)*addr + *size));
+ #elif defined __HAIKU__
+     thread_info thinfo;
+     CHECK_ERR(get_thread_info(find_thread(NULL), &thinfo));
+     *addr = thinfo.stack_end;
+     *size = (size_t)(thinfo.stack_end - thinfo.stack_base);
  #else
  #error STACKADDR_AVAILABLE is defined but not implemented.
  #endif
